water/lib/Dialect/Wave/Transforms/LowerReadWriteOps.cpp

```diff
@@ -125,6 +127,96 @@ materializeAffine(Location loc, ArrayRef<Attribute> symbols, AffineMap map,
+      scf::ForOp parentFor = nullptr;
+
+      // Look for a parent scf.for operation.
+      Operation *current = currentBlock->getParentOp();
+      while (current) {
+        if (auto forOp = dyn_cast<scf::ForOp>(current)) {
+          parentFor = forOp;
+          break;
+        }
+        current = current->getParentOp();
+      }
+
+      if (!parentFor) {
+        // Check if we're inside a wave.iterate instead.
+        Operation *currentOp = currentBlock->getParentOp();
+        while (currentOp) {
+          if (isa<wave::IterateOp>(currentOp)) {
+            return rewriter.notifyMatchFailure(
+                loc, "iteration symbol found inside wave.iterate - "
+                     "please run lower-wave-control-flow pass first");
+          }
+          currentOp = currentOp->getParentOp();
+        }
+
+        return rewriter.notifyMatchFailure(
+            loc, "iteration symbol found but no iteration context available");
+      }
+
+      // Get the induction variable and compute iteration value.
+      Value inductionVar = parentFor.getInductionVar();
+
+      // We need the tile size to compute iteration_value = induction_var *
+      // tile_size Get this from the function's tiling constraints.
+      func::FuncOp parentFunc = current->getParentOfType<func::FuncOp>();
+      if (!parentFunc) {
+        return rewriter.notifyMatchFailure(loc, "no parent function found");
+      }
+
+      auto hyperparamAttr =
+          parentFunc->getAttrOfType<wave::WaveHyperparameterAttr>(
+              wave::WaveDialect::kHyperparameterAttrName);
+      ArrayAttr constraints = parentFunc->getAttrOfType<ArrayAttr>(
+          wave::WaveDialect::kWaveConstraintsAttrName);
+
+      if (!hyperparamAttr || !constraints) {
+        return rewriter.notifyMatchFailure(
+            loc, "missing hyperparameters or constraints");
+      }
+
+      // Find the tile size for this iteration symbol.
+      std::optional<int64_t> tileSize;
+      for (Attribute constraintAttr : constraints) {
+        auto tilingConstraint =
+            dyn_cast<wave::TilingConstraintAttr>(constraintAttr);
+        if (!tilingConstraint)
+          continue;
+
+        if (tilingConstraint.getDim().getName() == iterSymbol.getName()) {
+          wave::WaveExprListAttr tileSizeAttr = tilingConstraint.getTileSize();
+          std::optional<SmallVector<int64_t>> evaluatedTileSize =
+              wave::evaluateMapWithHyperparams(tileSizeAttr.getMap(),
+                                               tileSizeAttr.getSymbols(),
+                                               hyperparamAttr);
+          if (evaluatedTileSize && evaluatedTileSize->size() == 1) {
+            tileSize = (*evaluatedTileSize)[0];
+            break;
+          }
+        }
+      }
+
+      if (!tileSize) {
+        return rewriter.notifyMatchFailure(loc,
+                                           "could not determine tile size");
+      }
```

Comment:
I think I'd differentiate two cases:
 - there is no tiling constraint for the loop, then it is simply not tiled, its "tile size" is equal to the trip count (unless something different happens with pywave);
 - we failed to evaluate the map: this is an error; ideally it is a user-visible error and not just failure-to-lower.